chasm/tree.go

```diff
@@ -153,6 +164,210 @@ func NewEmptyTree(
+	}
+	return nil
+}
+
+func fieldName(f reflect.StructField) string {
+	if tagName := f.Tag.Get(fieldNameTag); tagName != "" {
+		return tagName
+	}
+	return f.Name
+}
+
+func (n *Node) setSerializedValue(
+	nodePath []string,
+	serializedNode *persistencespb.ChasmNode,
+) {
+	if len(nodePath) == 0 {
+		n.serializedValue = serializedNode
+		return
+	}
+
+	childName := nodePath[0]
+	childNode, ok := n.children[childName]
+	if !ok {
+		childNode = newNode(n.nodeBase, n, childName)
+		n.children[childName] = childNode
+	}
+	childNode.setSerializedValue(nodePath[1:], serializedNode)
+}
+
+// deserialize initializes the node's value from its serializedValue.
+// If value is of component type, it initializes every chasm.Field of it and sets node field but not value field
+// i.e. it doesn't deserialize recursively and must be called on every node separately.
+func (n *Node) deserialize(
+	valueT reflect.Type,
+) error {
+	if err := validateType(valueT); err != nil {
+		return err
+	}
+
+	if n.parent == nil {
+		// Top level node is always component.
+		return n.deserializeComponentNode(valueT)
+	}
+
+	switch n.serializedValue.GetMetadata().GetAttributes().(type) {
+	case *persistencespb.ChasmNodeMetadata_ComponentAttributes:
+		return n.deserializeComponentNode(valueT)
+	case *persistencespb.ChasmNodeMetadata_DataAttributes:
+		return n.deserializeDataNode(valueT)
+	case *persistencespb.ChasmNodeMetadata_CollectionAttributes:
+		panic("not implemented")
+	case *persistencespb.ChasmNodeMetadata_PointerAttributes:
+		// TODO: return serviceerror.NewInternal(...) instead.
+	}
+	return nil
+}
+
+func (n *Node) deserializeComponentNode(
+	valueT reflect.Type,
+) error {
+	// TODO: use n.serializedValue.GetComponentAttributes().GetType() instead to support deserialization to interface.
+	valueV := reflect.New(valueT.Elem())
+	if n.serializedValue == nil {
+		// Nothing to deserialize. Create an empty value and return.
+		return n.setValue(valueV.Interface(), fieldTypeComponent)
+	}
+
+	protoMessageFound := false
+	for i := 0; i < valueT.Elem().NumField(); i++ {
+		fieldV := valueV.Elem().Field(i)
+		fieldT := fieldV.Type()
+
+		if fieldT == UnimplementedComponentT {
+			continue
+		}
+
+		if fieldT.AssignableTo(protoMessageT) {
+			if protoMessageFound {
+				return serviceerror.NewInternal("only one proto field allowed in component")
+			}
+			protoMessageFound = true
+
+			value, err := n.unmarshalProto(n.serializedValue.GetData(), fieldT)
+			if err != nil {
+				return err
+			}
+			fieldV.Set(value)
+			continue
+		}
+
+		// chasm.Field field must NOT be a pointer, i.e. chasm.Field[T] not *chasm.Field[T].
+		if fieldT.Kind() == reflect.Ptr {
+			continue
+		}
+
+		fieldN := fieldName(valueT.Elem().Field(i))
+
+		switch genericTypePrefix(fieldT) {
+		case chasmFieldTypePrefix:
+			if childNode, found := n.children[fieldN]; found {
+				// TODO: support chasm.Field[interface], type should go from registry
```

Comment:
My thinking is deserialize is dumb and always deserialize to a concrete value and cache it in the `value` field. The conversion to Interface can be done inside chasm.Field[Interface].Get() implementation. 

Either approach works for me and your's is probably cleaner and I'd like to see how Field.Get() is implemented.
I guess what I want to avoid though is duplicated decoding. If the decoded value is already available, then just return it, because for both component and data there's only one underlying *Struct type.